rust/connlib/tunnel/src/ice.rs

```diff
@@ -0,0 +1,111 @@
+use crate::PollNextIceCandidate;
+use connlib_shared::messages::{ClientId, GatewayId};
+use futures::channel::mpsc::Receiver;
+use futures_bounded::{PushError, StreamMap};
+use std::collections::HashMap;
+use std::task::{ready, Context, Poll};
+use std::time::Duration;
+use webrtc::ice_transport::ice_candidate::RTCIceCandidateInit;
+
+/// The [`Tunnel`]s ICE state on the client.
+///
+/// We split the receivers of ICE candidates into two phases because we only want to start sending them once we've received an SDP from the gateway.
+pub struct ClientIceState {
```

Comment:
Perhaps we should move these to their own crate(client/gateway) to be clearer on where they are used, even if that means including a new dependency on those crates?

Reply:
Well, we can't have circular dependencies between tunnel and client/gateway so I think for the moment, we have to keep these in the `firezone-tunnel` crate. But yeah, in an ideal world, these would live outside. I think the way to do this is to move all the specific state _into_ these structs. What I'd like to avoid is adding more functions to `RoleState` to access specific things.

I'll document this on the trait.